File: src/main/java/app/search/stages/SearchStage.java
package app.search.stages;

import app.search.controllers.SearchController;
import app.search.listcells.ResultListCell;
import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXMLLoader;
import javafx.geometry.Rectangle2D;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.ListCell;
import javafx.scene.control.ListView;
import javafx.scene.image.Image;
import javafx.scene.input.KeyCode;
import javafx.scene.paint.Color;
import javafx.stage.Screen;
import javafx.stage.Stage;
import javafx.stage.StageStyle;
import javafx.util.Callback;
import system.ResourceUtils;
import system.search.SearchEngine;
import system.search.results.*;
import utils.ImageResolver;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public class SearchStage extends Stage {
    public static final int MAX_RESULTS = 6; // Maximum number of results
    public static final int MAX_RESULTS_FOR_AGENT = 3; // Maximum number of results for each agent

    private SearchController controller;
    private ResourceBundle resourceBundle;
    private SearchEngine searchEngine;

    private int screenCenterX;
    private int screenCenterY;

    /**
     * This map is used to store image caches
     */
    private ConcurrentHashMap<String, Image> imageCacheMap = new ConcurrentHashMap<>();

    // In this list are registered the priority of the results in the search bar
    // The first elements are displayed first
    private List<Class<? extends AbstractResult>> resultPriorityList = new ArrayList<>(20);

    // Similar to the resultPriorityList, here are registered only the types of filter
    // the user can select. For example, the terminal filter cannot be selected by the user.
    private List<Class<? extends AbstractResult>> userFilters = new ArrayList<>(10);

    // If this is set, only show the results of this category
    private Class<? extends AbstractResult> resultFilter = null;

    // The result list, when modified the listview updates
    private ObservableList<AbstractResult> observableResults = FXCollections.observableArrayList();

    // The current results for the given query
    private Map<Class<? extends AbstractResult>, List<? extends AbstractResult>> currentResults;

    public SearchStage(ResourceBundle resourceBundle, SearchEngine searchEngine) throws IOException {
        this.resourceBundle = resourceBundle;
        this.searchEngine = searchEngine;
        FXMLLoader fxmlLoader = new FXMLLoader(ResourceUtils.getResource("/layouts/search_dialog.fxml").toURI().toURL());
        fxmlLoader.setResources(resourceBundle);
        Parent root = fxmlLoader.load();
        Scene scene = new Scene(root);
        scene.getStylesheets().add(ResourceUtils.getResource("/css/search_dialog.css").toURI().toString());
        this.setTitle("Dokey Search");  // DON'T CHANGE, used on window in the focus mechanism.
        this.setScene(scene);
        this.setAlwaysOnTop(true);
        initStyle(StageStyle.TRANSPARENT);
        scene.setFill(Color.TRANSPARENT);
        this.getIcons().add(new Image(SearchStage.class.getResourceAsStream("/assets/icon.png")));

        controller = (SearchController) fxmlLoader.getController();
        controller.resultListView.setManaged(false);

        // Hide the filter box
        controller.filterBox.setManaged(true);
        controller.filterBox.setVisible(true);

        // Setup the list cells
        Image fallback = ImageResolver.getInstance().getImage(SearchStage.class.getResourceAsStream("/assets/photo.png"), 32);
        controller.resultListView.setCellFactory(new Callback<ListView<AbstractResult>, ListCell<AbstractResult>>() {
            @Override
            public ListCell<AbstractResult> call(ListView<AbstractResult> param) {
                return new ResultListCell(resourceBundle, fallback, imageCacheMap);
            }
        });

        // Setup the text field search callbacks
        controller.queryTextField.textProperty().addListener((observable, oldValue, newValue) -> {
            searchEngine.requestQuery(newValue.trim(), (results) -> {
                currentResults = results;

                renderResults();
            });
        });
        // If someone deselects the textfield, re select it
        controller.queryTextField.focusedProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue == false) {
                Platform.runLater(() -> controller.queryTextField.requestFocus());
            }
        });

        // Setup keyboard events
        controller.queryTextField.setOnKeyPressed(event -> {
            int currentlySelected = controller.resultListView.getSelectionModel().getSelectedIndex();
            if (event.getCode() == KeyCode.DOWN) {  // Select next element in the list
                if (currentlySelected < (controller.resultListView.getItems().size() - 1)) {
                    controller.resultListView.getSelectionModel().select(currentlySelected + 1);
                }else{  // Select the first
                    if (controller.resultListView.getItems().size() > 0) {
                        controller.resultListView.getSelectionModel().select(0);
                    }
                }

                event.consume();
            } else if (event.getCode() == KeyCode.UP) {  // Select previous element in the list
                if (currentlySelected > 0) {
                    controller.resultListView.getSelectionModel().select(currentlySelected - 1);
                }else{  // Select the last one
                    if (controller.resultListView.getItems().size() > 0) {
                        controller.resultListView.getSelectionModel().select(controller.resultListView.getItems().size()-1);
                    }
                }

                event.consume();
            } else if (event.getCode() == KeyCode.ENTER) {  // Execute action and close the stage
                AbstractResult result = (AbstractResult) controller.resultListView.getSelectionModel().getSelectedItem();
                executeSearch(result);
            } else if (event.getCode() == KeyCode.ESCAPE) { // Close the search stage or remove filter
                if (resultFilter != null) {  // REMOVE FILTER
                    resultFilter = null;
                    renderResults();
                }else{  // CLOSE
                    close();
                }
            } else if (event.getCode() == KeyCode.TAB) {  // Filter results
                if ((controller.resultListView.getSelectionModel().getSelectedItem() == null) ||
                        ((AbstractResult) controller.resultListView.getSelectionModel().getSelectedItem())
                                .getClass().equals(resultFilter)) {
                    int index = 0;
                    if (resultFilter != null) {
                        int nextIndex = userFilters.indexOf(resultFilter) + 1;
                        if (nextIndex != -1) {
                            if (nextIndex >= userFilters.size()) {
                                index = 0;
                            }else{
                                index = nextIndex;
                            }
                        }
                    }
                    resultFilter = userFilters.get(index);
                    renderResults();
                }else{
                    if (controller.resultListView.getSelectionModel().getSelectedItem() != null) {
                        Class<? extends AbstractResult> selectedClass = (Class<? extends AbstractResult>) controller.resultListView.getSelectionModel().getSelectedItem().getClass();
                        // Make sure the selected item is filterable from the user
                        // for example, the terminal cannot be filtered
                        if (userFilters.contains(selectedClass)) {
                            resultFilter = selectedClass;
                            renderResults();
                        }
                    }
                }

                event.consume();
            }
        });
        // Detect if window lose focus
        focusedProperty().addListener(((observable, oldValue, newValue) -> {
            if (newValue == false)
                close();
        }));

        // Result list view click listener, execute the corresponding action
        controller.resultListView.setOnMouseClicked(event -> {
            AbstractResult result = (AbstractResult) controller.resultListView.getSelectionModel().getSelectedItem();
            executeSearch(result);
        });

        registerResultPriorityList();
        registerUserFilterList();

        controller.resultListView.setItems(observableResults);

        preInitialize();
    }

    /**
     * Actions that must be done before showing the bar to initialize it to its empty state.
     */
    public void preInitialize() {
        // Reset the result data structures
        currentResults = new HashMap<>();
        observableResults.clear();

        resultFilter = null;

        // Reset the query text field to an empty string
        controller.queryTextField.setText("");

        renderResults();
    }

    /**
     * Actions that must be done after showing the bar to initialize it.
     */
    public void postInitialize() {
        Platform.runLater(() -> positionBarOnScreen());
    }

    /**
     * Position the bar in the correct position of the screen
     */
    private void positionBarOnScreen() {
        // Calculate the correct coordinates for the center of the screen
        Rectangle2D primScreenBounds = Screen.getPrimary().getVisualBounds();
        setX((primScreenBounds.getWidth() - getWidth()) / 2);
        setY((primScreenBounds.getHeight() - getHeight()) / 4 * 1);
    }

    private void renderResults() {
        // Render filter label
        Platform.runLater(() -> {
            // Show the filter label
            if (resultFilter != null) {
                // Get the filter text
                try {
                    String labelID = (String) resultFilter.getDeclaredField("SEARCH_FILDER_RESOURCE_ID").get(null);
                    String filterText = resourceBundle.getString(labelID);
                    if (filterText != null) {
                        controller.filterLabel.setText(filterText);
                        controller.filterBox.setManaged(true);
                        controller.filterBox.setVisible(true);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }else{  // hide the filter label
                controller.filterBox.setManaged(false);
                controller.filterBox.setVisible(false);
            }
        });

        if (currentResults == null)
            return;

        ArrayList<AbstractResult> priorityResults = new ArrayList<>(20);

        // If the results has more than one category, limit the results for each one
        boolean limitResultsForCategory = currentResults.keySet().size() > 1;

        // If there is a result filter, only show those results
        if (resultFilter != null) {
            if (currentResults.containsKey(resultFilter)) {
                for (AbstractResult result : currentResults.get(resultFilter)) {
                    if (priorityResults.size() < MAX_RESULTS) {
                        priorityResults.add(result);
                    }
                }
            }
        }else{  // No filter, show all
            // Get the result for each result category
            for (Class<? extends AbstractResult> resultClass : resultPriorityList) {
                if (currentResults.containsKey(resultClass)) {
                    int currentResult = 0;

                    for (AbstractResult result : currentResults.get(resultClass)) {
                        if ((currentResult < MAX_RESULTS_FOR_AGENT || !limitResultsForCategory)
                                && priorityResults.size() < MAX_RESULTS) {
                            priorityResults.add(result);
                            currentResult++;
                        }else{
                            break;
                        }
                    }
                }
            }
        }

        // Update the list view
        Platform.runLater(() -> {
            observableResults.setAll(priorityResults);

            // Select the first item
            if (priorityResults.size() > 0) {
                controller.resultListView.getSelectionModel().select(0);
            }

            // Set the height based on the list view
            controller.resultListView.setPrefHeight(priorityResults.size() * ResultListCell.ROW_HEIGHT);

            // Show the list view and refresh stage size to fit all contents
            controller.resultListView.setManaged(true);
            sizeToScene();
        });
    }

    /**
     * Register the result type priority in the list
     */
    private void registerResultPriorityList() {
        resultPriorityList.add(QuickCommandResult.class);
        resultPriorityList.add(AddUrlToQuickCommandsResult.class);
        resultPriorityList.add(CalculatorResult.class);
        resultPriorityList.add(TerminalResult.class);
        resultPriorityList.add(DebugResult.class);
        resultPriorityList.add(ApplicationResult.class);
        resultPriorityList.add(ShortcutResult.class);
        resultPriorityList.add(BookmarkResult.class);
        resultPriorityList.add(GoogleSearchResult.class);
    }

    /**
     * Register the user filters.
     */
    private void registerUserFilterList() {
        userFilters.add(GoogleSearchResult.class);
        userFilters.add(BookmarkResult.class);
        userFilters.add(ShortcutResult.class);
        userFilters.add(ApplicationResult.class);
        userFilters.add(QuickCommandResult.class);
    }

    private void executeSearch(AbstractResult result) {
        if (result != null) {
            new Thread(() -> {
                result.executeAction();
            }).start();
            close();
        }
    }
}

File: src/main/java/app/search/stages/SearchStage.kt
package app.search.stages

import app.search.controllers.SearchController
import app.search.sectionlistview.SectionListView
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.subjects.PublishSubject
import javafx.application.Platform
import javafx.fxml.FXMLLoader
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.image.Image
import javafx.scene.input.KeyCode
import javafx.scene.paint.Color
import javafx.stage.Screen
import javafx.stage.Stage
import javafx.stage.StageStyle
import org.reflections.Reflections
import system.ResourceUtils
import system.applications.Application
import system.applications.ApplicationManager
import system.image.ImageResolver
import system.search.SearchEngine
import system.search.annotations.FilterableResult
import system.search.results.Result
import system.search.results.ResultCategory
import utils.OSValidator
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.reflect.KClass

/**
 * The Dokey Search stage.
 */
class SearchStage @Throws(IOException::class)
constructor(private val resourceBundle: ResourceBundle, private val searchEngine: SearchEngine,
            private val imageResolver: ImageResolver) : Stage() {

    private val controller: SearchController

    // This list view will hold the results of the search
    private val listView : SectionListView

    // Here are registered only the types of filter the user can select.
    // For example, the terminal filter cannot be selected by the user.
    private val filterableResults : MutableSet<KClass<out Result>> = mutableSetOf()

    // If this is set, only show the results of this category
    private var resultFilter: ResultCategory? = null

    // The current results for the given query
    private var currentResults: MutableList<Result>? = null

    // This subject is used to debounce the display of results, to avoid flickering
    private val resultSubject = PublishSubject.create<SortedMap<ResultCategory, MutableList<Result>>>()

    // Current search query
    private var currentQuery: String? = null

    // The application that was active before the bar was shown
    private var activeApplication: Application? = null

    init {
        val fxmlLoader = FXMLLoader(ResourceUtils.getResource("/layouts/search_dialog.fxml")!!.toURI().toURL())
        fxmlLoader.resources = resourceBundle
        val root = fxmlLoader.load<Parent>()
        val scene = Scene(root)
        scene.stylesheets.add(ResourceUtils.getResource("/css/search_dialog.css")!!.toURI().toString())
        this.title = "Dokey Search"  // DON'T CHANGE, used on window in the focus mechanism.
        this.scene = scene
        this.isAlwaysOnTop = true
        initStyle(StageStyle.TRANSPARENT)
        scene.fill = Color.TRANSPARENT
        this.icons.add(Image(SearchStage::class.java.getResourceAsStream("/assets/icon.png")))

        controller = fxmlLoader.getController<Any>() as SearchController

        // Setup the bar width
        controller.rootNode.prefWidth = DIALOG_WIDTH

        // Setup the listview
        listView = SectionListView(imageResolver) {
            this.hide()  // Hide the window when a user clicks on a result
        }

        controller.rootNode.children.add(listView)
        listView.isManaged = false

        // Hide the filter box
        controller.filterBox.isManaged = true
        controller.filterBox.isVisible = true

        // Setup the text field search callbacks
        controller.queryTextField.textProperty().addListener { _, _, searchQuery ->
            currentQuery = searchQuery

            // Reset the current result set
            currentResults?.clear()
            elaborateResults()

            if (!searchQuery.isBlank()) {
                searchEngine.requestQuery(searchQuery.trim(), activeApplication) { query, results ->
                    // Make sure the result is relative to the current query
                    if (currentQuery == query) {
                        currentResults!!.addAll(results)
                        elaborateResults()
                    }
                }
            }
        }

        // If someone deselects the textfield, re select it
        controller.queryTextField.focusedProperty().addListener { _, _, isFocused ->
            if (!isFocused) {
                Platform.runLater { controller.queryTextField.requestFocus() }
            }
        }

        // Setup keyboard events
        controller.queryTextField.setOnKeyPressed { event ->
            val currentlySelected = listView.getSelectedIndex()
            if (event.code == KeyCode.DOWN) {  // Select next element in the list
                if (currentlySelected < listView.getTotalItems() - 1) {
                    listView.selectIndex(listView.getSelectedIndex() + 1)
                } else {  // Select the first
                    if (listView.getTotalItems() > 0) {
                        listView.selectIndex(0)
                    }
                }

                event.consume()
            } else if (event.code == KeyCode.UP) {  // Select previous element in the list
                if (currentlySelected > 0) {
                    listView.selectIndex(listView.getSelectedIndex() - 1)
                } else {  // Select the last one
                    if (listView.getTotalItems() > 0) {
                        listView.selectIndex(listView.getTotalItems() - 1)
                    }
                }

                event.consume()
            }
            else if (event.code == KeyCode.ENTER) {  // Execute action and close the stage
                listView.executeCurrentResult()
                this.hide()
            } else if (event.code == KeyCode.ESCAPE) { // Close the search stage or remove filter
                if (resultFilter != null) {  // REMOVE FILTER
                    resultFilter = null
                    elaborateResults()
                } else {  // CLOSE
                    closeAndRefocus()
                }
            } else if (event.code == KeyCode.ALT) { // Filter
                val currentResult = listView.getSelectedResult()

                if (currentResult != null) {
                    if (resultFilter == null && filterableResults.contains(currentResult::class)) {
                        resultFilter = currentResult.category
                    }

                    elaborateResults()
                }
            } else if (event.code == KeyCode.TAB) {  // Select next category
                listView.selectNextSection()

                event.consume()
            }
        }
        // Detect if window lose focus
        focusedProperty().addListener { _, _, isFocused ->
            if (isFocused == false)
                closeAndRefocus()
        }

        registerFilterableResults()

        // Setup the debouncing mechanism to avoid flickering when displaying the results
        resultSubject.debounce(50, TimeUnit.MILLISECONDS)
                .subscribe(object : Observer<SortedMap<ResultCategory, MutableList<Result>>> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onNext(results: SortedMap<ResultCategory, MutableList<Result>>) {
                // Update the list view
                Platform.runLater {
                    // Render filter label
                    if (resultFilter != null) {
                        // Get the filter text
                        try {
                            controller.filterLabel.text = resultFilter?.name
                            controller.filterBox.isManaged = true
                            controller.filterBox.isVisible = true
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }

                    } else {  // hide the filter label
                        controller.filterBox.isManaged = false
                        controller.filterBox.isVisible = false
                    }

                    // Render results
                    val emptySearch = (currentQuery == null) || (currentQuery != null && currentQuery!!.isBlank())
                    if (results.isNotEmpty() || emptySearch) {
                        listView.setResults(results)

                        // Select the first item
                        if (results.size > 0) {
                            listView.selectIndex(0)
                        }

                        // Set the height based on the list view
                        listView.adaptHeight()

                        // Show the list view and refresh stage size to fit all contents
                        listView.isManaged = true
                        sizeToScene()
                    }
                }
            }

            override fun onError(e: Throwable) {

            }

            override fun onComplete() {

            }
        })
    }

    /**
     * Actions that must be done before showing the bar to initialize it to its empty state.
     */
    fun preInitialize(activeApplication: Application?) {
        // Reset the result data structures
        currentResults = mutableListOf()

        resultFilter = null

        // Reset the query text field to an empty string
        controller.queryTextField.text = ""

        this.activeApplication = activeApplication

        elaborateResults()
    }

    /**
     * Actions that must be done after showing the bar to initialize it.
     */
    fun postInitialize() {
        // Reset the height of the list box so that it contains no elements
        listView.prefHeight = 0.0

        // Resize the scene to fit the listview
        sizeToScene()

        // Center the bar in the screen
        positionBarOnScreen()
    }

    /**
     * Position the bar in the correct position of the screen
     */
    private fun positionBarOnScreen() {
        // Calculate the correct coordinates for the center of the screen
        val primScreenBounds = Screen.getPrimary().visualBounds
        x = (primScreenBounds.width - width) / 2
        y = (primScreenBounds.height - height) / 6 * 1
    }

    /**
     * Elaborate the currentResults by grouping them into categories, ordering them
     * based on the priority and trimming to the correct number of results per category.
     */
    private fun elaborateResults() {
        if (currentResults == null)
            return

        if (currentResults!!.isEmpty()) {
            // Signal the empty list
            resultSubject.onNext(TreeMap<ResultCategory, MutableList<Result>>())
        }

        // Add all the current results to the initial set of results
        var initialResults = mutableListOf<Result>()
        initialResults.addAll(currentResults!!)

        // The maximum number of results for each agent, this will be recalculated later
        var maxNumberOfResultsForAgent = MAX_RESULTS

        if (resultFilter != null) {
            // Filter the initial results based on the filtered category
            initialResults = initialResults.filter { it.category == resultFilter }.toMutableList()
        }else{
            // Calculate the number of maximum number of results for each agent dynamically

            // Get the total number of categories to calculate the number of results for each category
            val totalNumberOfCategories = initialResults.map { it.category.name }.distinct().count()

            // Clip the number of category to the MAX_DISPLAYED_CATEGORIES
            val numberOfCategories = if (totalNumberOfCategories > MAX_DISPLAYED_CATEGORIES)
                MAX_DISPLAYED_CATEGORIES
            else
                totalNumberOfCategories

            // Calculate the maximum number of results based on the number of categories
            maxNumberOfResultsForAgent = if (numberOfCategories > 0 )
                MAX_RESULTS / numberOfCategories
            else
                MAX_RESULTS
        }

        // This sorted map will hold the prioritized results
        val finalResults : SortedMap<ResultCategory, MutableList<Result>> = TreeMap<ResultCategory, MutableList<Result>>()

        var totalResultCount = 0

        for (result in initialResults) {
            // Stop the process if the maximum number of results is reached
            if (totalResultCount > MAX_RESULTS) {
                break
            }

            // Initialize the list for the given category
            if (finalResults[result.category] == null) {
                finalResults[result.category] = mutableListOf()
            }else{
                // Clip the number of results for the given agent if exceeded
                if (finalResults[result.category]!!.size > maxNumberOfResultsForAgent) {
                    continue
                }
            }

            finalResults[result.category]!!.add(result)
            totalResultCount++
        }

        // Signal the new results
        resultSubject.onNext(finalResults)
    }

    /**
     * Register the user filters.
     */
    private fun registerFilterableResults() {
        // Load all the filterable result
        val reflections = Reflections("system.search.results")
        val resultClasses = reflections.getTypesAnnotatedWith(FilterableResult::class.java)
        resultClasses.forEach { resultClass ->
            filterableResults.add((resultClass as Class<out Result>).kotlin)
        }
    }

    /**
     * In some systems, after the window has focused, the bar should focus back to the original
     * application ( like in MacOS ), because the OS does not do that automatically.
     */
    private fun closeAndRefocus() {
        if (OSValidator.isMac()) {
            Platform.runLater { activeApplication?.open() }
        }
        close()
    }

    companion object {
        val MAX_RESULTS = 9 // Maximum number of results
        val MAX_DISPLAYED_CATEGORIES = 4  // Maximum number of categories shown at once

        val DIALOG_WIDTH = 700.0  // The width of the BAR
    }
}